tool/src/js/type_generation/mod.rs

```diff
@@ -217,35 +215,57 @@ impl<'ctx, 'tcx> TyGenContext<'ctx, 'tcx> {
                 None
             };
 
-            let field_layout = crate::js::layout::type_size_alignment(&field.ty, self.tcx);
+            let padding = struct_field_info.fields[i].padding_count;
 
-            let curr_offset = offsets[i];
-            let next_offset = if i < offsets.len() - 1 {
-                offsets[i + 1]
-            } else {
-                curr_offset + field_layout.size()
-            };
 
-            let padding = next_offset - curr_offset - field_layout.size();
+            let force_padding = match (struct_field_info.fields[i].scalar_count, struct_field_info.scalar_count) {
+                // There's no padding needed
+                (0 | 1, _) => ForcePaddingStatus::NoForce,
+                // Non-structs don't care
+                _ if !matches!(&field.ty, &hir::Type::Struct(_)) => ForcePaddingStatus::NoForce,
+                // 2-field struct contained in 2-field struct, caller decides
+                (2, 2) => {
+                    needs_force_padding = true;
+                    ForcePaddingStatus::PassThrough
+                }
+                // Outer struct has > 3 fields, always pad
+                (2, 3..) => ForcePaddingStatus::Force,
+                // Larger fields will always have padding anyway
+                _ => ForcePaddingStatus::NoForce
```

Comment:
Thought: Wow, very specific rules. I trust this is the result of your research

Reply:
Yeah